src/webots/nodes/WbBaseNode.cpp

```diff
@@ -124,6 +125,269 @@ void WbBaseNode::validateProtoNodes() {
+  for (int i = 0; i < internalProtoNodes.size(); ++i) {
+    internalProtoNodes[i]->disconnectInternalNode();
+    const QVector<WbField *> fields = internalProtoNodes[i]->fields();
+
+    for (int j = 0; j < fields.size(); j++)
+      fields[j]->setParameter(NULL);
+
+    internalProtoNodes[i]->setProtoParameterNode(NULL);  // break link with proto parameter node
+  }
+
+  // break link [parameter] -> [internal field] and [parameter node] -> [internal node] (from parameter node side)
+  for (int i = 0; i < invisibleProtoParameterNodes.size(); ++i) {
+    invisibleProtoParameterNodes[i]->clearProtoParameterNodeInstances();  // clear downward references
+
+    // clear internal field references (for protoParameterNodes the reference is kept in its fields)
+    QVector<WbField *> fields = invisibleProtoParameterNodes[i]->fields();
+    for (int j = 0; j < fields.size(); ++j)
+      fields[j]->clearInternalFields();
+  }
+
+  // now the proto parameter nodes can be deleted, depending on the situation it can either be in the parameter or field side of
+  // the parent node. The signal is not emitted to prevent the internal node from being deleted as well in the process
+  for (int i = 0; i < invisibleProtoParameterNodes.size(); ++i) {
+    WbNode *parameterNode = invisibleProtoParameterNodes[i];
+    WbNode *parent = parameterNode->parentNode();
+
+    const QVector<WbField *> fields = parent->fields();
+    for (int j = 0; j < fields.size(); ++j) {
+      WbSFNode *sfnode = dynamic_cast<WbSFNode *>(fields[j]->value());
+      WbMFNode *mfnode = dynamic_cast<WbMFNode *>(fields[j]->value());
+      if (sfnode && sfnode->value() == parameterNode) {
+        sfnode->blockSignals(true);
+        sfnode->setValue(NULL);
+        sfnode->blockSignals(false);
+        parent->removeFromFieldsOrParameters(fields[j]);
+      } else {
+        if (mfnode && mfnode->nodeIndex(parameterNode) != -1) {
+          mfnode->blockSignals(true);
+          mfnode->removeNode(parameterNode);
+          mfnode->blockSignals(false);
+          parent->removeFromFieldsOrParameters(fields[j]);
+        }
+      }
+    }
+
+    const QVector<WbField *> parameters = parent->parameters();
+    for (int j = 0; j < parameters.size(); j++) {
+      WbSFNode *sfnode = dynamic_cast<WbSFNode *>(parameters[j]->value());
+      WbMFNode *mfnode = dynamic_cast<WbMFNode *>(parameters[j]->value());
+
+      if (sfnode && sfnode->value() == parameterNode) {
+        sfnode->blockSignals(true);
+        sfnode->setValue(NULL);
+        sfnode->blockSignals(false);
+        parent->removeFromFieldsOrParameters(parameters[j]);
+      } else {
+        if (mfnode && mfnode->nodeIndex(parameterNode) != -1) {
+          mfnode->blockSignals(true);
+          mfnode->removeNode(parameterNode);
+          mfnode->blockSignals(false);
+          parent->removeFromFieldsOrParameters(parameters[j]);
+        }
+      }
+    }
```

Comment:
Instead of duplicating this block you should merge fields and parameters in the same QVector and run the loop only once.